Challenges workflow and business logic implementation

lankastersky opened the pull request.

lankastersky committed:
Challenges workflow and business logic implementation

lankastersky committed:
Fixed issue with decoding challenges

lankastersky committed:
Storing challenges in ChallengesModel instead of ChallengesProvider

lankastersky committed:
Business logic of challenges model

NikAshanin commented on zen/models/Challenge.swift:
```diff
@@ -1,9 +1,90 @@
+    let prevFinishedTimes: [TimeInterval]? = nil
+    let prevRatings: [Float]? = nil
+
+    mutating func updateStatus() {
```
делай класс, никаких mutation func

lankastersky replied:
done

NikAshanin commented on zen/models/Challenge.swift:
```diff
@@ -1,9 +1,90 @@
+        }
+    }
+
+    mutating func decline() {
```
тоже самое

lankastersky replied:
мне нужна защита против обработки оставшихся статусов

NikAshanin commented on zen/models/Challenge.swift:
```diff
@@ -1,9 +1,90 @@
+        }
+    }
+
+    mutating func reset() {
```
тоже самое

lankastersky replied:
done

NikAshanin commented on zen/models/Challenge.swift:
```diff
@@ -1,9 +1,90 @@
+        case .accepted?:
+            status = .finished
+            finishedTime = Date().timeIntervalSince1970
+        default:
```
вроде можно без default

lankastersky replied:
мне нужна защита против обработки оставшихся статусов

NikAshanin replied:
лучше так и напиши, что такие статусы неприемлимы. Просто если default не будет, а ты расширишь этот enum, то сразу увидишь где нужно что добавить. С default так не сработает и будут ошибки, которые нужно руками искать

lankastersky replied:
done

NikAshanin commented on zen/models/Challenge.swift:
```diff
@@ -1,9 +1,90 @@
+        case .shown?, .accepted?:
+            status = .declined
+            finishedTime = Date().timeIntervalSince1970
+        default:
```
вроде можно без default

lankastersky replied:
мне нужна защита против обработки оставшихся статусов

NikAshanin replied:
лучше так и напиши, что такие статусы неприемлимы. Просто если default не будет, а ты расширишь этот enum, то сразу увидишь где нужно что добавить. С default так не сработает и будут ошибки, которые нужно руками искать

lankastersky replied:
done

NikAshanin commented on zen/models/Challenge.swift:
```diff
@@ -1,9 +1,90 @@
+        if let decodedId = decoder.codingPath.first?.stringValue {
+            challengeId = decodedId
+        } else {
+            print("Challenge id is not defined")
```
может быть такое, что нет ID? я бы сделал падение в этом месте

lankastersky replied:
сделал падение

NikAshanin commented on zen/models/Challenge.swift:
```diff
@@ -1,9 +1,90 @@
+            challengeId = decodedId
+        } else {
+            print("Challenge id is not defined")
+            challengeId = ""
```
нене, вот это что-то выглядит плохо

lankastersky replied:
согласен

NikAshanin commented on zen/models/ChallengeData.swift:
```diff
@@ -0,0 +1,13 @@
+import Foundation
+
+/// Modifiable challenge data for serialization etc.
+struct ChallengeData {
```
никаких структур. структуры в swift используй только как это делает apple (CGPoint, CGRect и тд)

lankastersky replied:
done

NikAshanin commented on zen/models/ChallengesModel.swift:
```diff
@@ -0,0 +1,326 @@
+    private static let proportionAcceptHighLevelChallenge = 2.0 / 3
+
+    private var challengesMap: [String: Challenge] = [:]
+    private var currentChallengeId = ""
```
wtf?)

lankastersky replied:
он исползуется как ключ в словаре

NikAshanin replied:
пустой строкой быть не может

NikAshanin commented on zen/models/ChallengesModel.swift:
```diff
@@ -0,0 +1,326 @@
+    private static let proportionAcceptMediumLavelChallenge = 1.0 / 3
+    private static let proportionAcceptHighLevelChallenge = 2.0 / 3
+
+    private var challengesMap: [String: Challenge] = [:]
```
что за MAP? Dictionary это, нет MAP в swift

lankastersky replied:
хаха да) но мне так привычней

NikAshanin replied:
Тебе то да, но ты пишешь в том числе не для себя. И надо следовать стандартам, которые приняты в языке, который ты используешь, чтобы было понятно ДРУГИМ людям :)

NikAshanin commented on zen/models/ChallengesModel.swift:
```diff
@@ -0,0 +1,326 @@
+    private var challengesMap: [String: Challenge] = [:]
+    private var currentChallengeId = ""
+    private var currentChallengeShownTime: TimeInterval = 0
+    private(set) var currentLevel: ChallengeLevel = ChallengeLevel.low
```
private(set) var currentLevel: ChallengeLevel = .low

lankastersky replied:
done

NikAshanin commented on zen/models/ChallengesModel.swift:
```diff
@@ -0,0 +1,326 @@
+    private let challengesArchiver = ChallengesArchiver()
+
+    var currentChallenge: Challenge? {
+        get { return challengesMap[currentChallengeId] }
```
возможно падение же. небезопасный код

lankastersky replied:
почему падение? опциональный тип возвращается

NikAshanin commented on zen/models/ChallengesModel.swift:
```diff
@@ -0,0 +1,326 @@
+
+    var shownChallengesNumber: Int {
+        get {
+            let shownNumber: Int = challengesByStatus(ChallengeStatus.shown).count
```
Int зачем писать это? count вернет только инт)

lankastersky replied:
done

NikAshanin commented on zen/models/ChallengesModel.swift:
```diff
@@ -0,0 +1,326 @@
+    var sortedChallenges: [Challenge] {
+        get {
+            let challenges = challengesMap.values
+            return challenges.sorted(by: { (left, right) -> Bool in
```
return challenges.sorted(by: { $0. challengeId < $1. challengeId })

lankastersky replied:
done

NikAshanin commented on zen/models/ChallengesModel.swift:
```diff
@@ -0,0 +1,326 @@
+    }
+
+    func finishedChallengesSortedByTime() -> [Challenge] {
+        return finishedChallenges.sorted(by: { (left, right) -> Bool in
```
тоже самое

lankastersky replied:
done

NikAshanin commented on zen/models/ChallengesModel.swift:
```diff
@@ -0,0 +1,326 @@
+    }
+
+    func finishedChallengesSortedByTimeDesc() -> [Challenge] {
+        return finishedChallenges.sorted(by: { (left, right) -> Bool in
```
тоже самое

NikAshanin commented on zen/models/ChallengesModel.swift:
```diff
@@ -0,0 +1,326 @@
+
+    /// Persistently stores challenges and refreshes challenges map
+    func storeChallenges(_ challengesMap: [String: Challenge]) {
+        self.challengesMap = challengesMap
```
self redundant...

lankastersky replied:
done

NikAshanin commented on zen/models/ChallengesModel.swift:
```diff
@@ -0,0 +1,326 @@
+        if challengesMap.isEmpty {
+            if let challenges = challengesArchiver.restoreChallenges() {
+                for challenge in challenges {
+                    challengesMap[challenge.challengeId] = challenge
```
тут падение

lankastersky replied:
можешь пояснить?

NikAshanin commented on zen/models/ChallengesModel.swift:
```diff
@@ -0,0 +1,326 @@
+
+    /// Restores challenges from persistent storage and selects current challenge if needed
+    func restoreChallenges() {
+        if challengesMap.isEmpty {
```
слишком высокая цикломатическая сложность. юзай гварды

lankastersky replied:
пробовал, пишет
'guard' body must not fall through, consider using a 'return' or 'throw' to exit the scope

NikAshanin replied:
а что не так? внутри должен быть return

NikAshanin commented on zen/models/ChallengesModel.swift:
```diff
@@ -0,0 +1,326 @@
+    func restoreChallenges() {
+        if challengesMap.isEmpty {
+            if let challenges = challengesArchiver.restoreChallenges() {
+                for challenge in challenges {
```
challenges.forEach

старайся юзать функциональный подход вообще

lankastersky replied:
done

NikAshanin commented on zen/models/ChallengesModel.swift:
```diff
@@ -0,0 +1,326 @@
+    }
+
+    private func storeState() {
+        challengesArchiver.storeChallengeData(challengesMap)
```
нельзя это всё вынести в challengesArchiver?

lankastersky replied:
наружу торчит только модель, она фасад

NikAshanin commented on zen/models/ChallengesModel.swift:
```diff
@@ -0,0 +1,326 @@
+
+        var newChallengeRequired = false
+
+        if currentChallengeId == "" {
```
и тут я долго пытался выразиться культурно :) В общем, ненадо так :)

lankastersky replied:
а как?

NikAshanin replied:
ну не может быть логически ID чего-то пустой строкой
тогда уж nil

NikAshanin commented on zen/models/ChallengesModel.swift:
```diff
@@ -0,0 +1,326 @@
+    private func selectChallengeIfNeeded() {
+        restoreState()
+
+        var newChallengeRequired = false
```
тут явно без этого можно красиво сделать, давай подумаем как это сделать красиво

lankastersky replied:
я уже думал)

NikAshanin commented on zen/models/ChallengesModel.swift:
```diff
@@ -0,0 +1,326 @@
+    /// Else get random challenge from all not equal to previous one.
+    ///
+    /// - Returns: new challenge Id
+    private func newChallengeId() -> String {
```
private var

lankastersky replied:
done

NikAshanin commented on zen/models/ChallengesModel.swift:
```diff
@@ -0,0 +1,326 @@
+            + challengesByStatus(.shown)
+        let filteredNonFinishedChallenges = challengesOfCurrentLevel(nonFinishedChallenges)
+
+        if !(filteredNonFinishedChallenges.isEmpty) {
```
минутка иронии: если цикломатическая сложность метода ниже 10, то он просто не пройдет ревью в гугл? :)
Давай упростим метод и разнесем в разные или как-то еще. условие, внутри условие, внутри условие и там еще цикл - это больно

lankastersky replied:
хаха ты прав) вот что бывает когда пишешь код по ночам) уменьшил сложность

NikAshanin commented on zen/models/ChallengesModel.swift:
```diff
@@ -0,0 +1,326 @@
+    /// - Returns: challenges of current and lower levels
+    private func challengesOfCurrentLevel(_ nonFinishedChallenges: [Challenge]) -> [Challenge] {
+        let challenges = Array(challengesMap.values)
+        if challenges.isEmpty {
```
guard

lankastersky replied:
как?

NikAshanin replied:
guard !challenges.isEmpty else { return [] }

lankastersky replied:
done

NikAshanin commented on zen/models/ChallengesModel.swift:
```diff
@@ -0,0 +1,326 @@
+        return challenges
+    }
+
+    private class func randomChallenge(_ challenges: [Challenge]) -> Challenge {
```
static

lankastersky replied:
done

NikAshanin commented on zen/models/ChallengesModel.swift:
```diff
@@ -0,0 +1,326 @@
+
+    private func challengesByStatus(_ challengeStatus: ChallengeStatus?) -> [Challenge] {
+        var challenges: [Challenge] = []
+        for challenge in challengesMap.values where challenge.status == challengeStatus {
```
try to use .filter
MORE functional style plz

lankastersky replied:
done

lankastersky committed:
Review fixes

NikAshanin commented on zen/models/Challenge.swift:
```diff
@@ -1,9 +1,86 @@
+import Foundation
+
 /// Challenge model
```
Вот с именованием бы поработать прям. Я хз откуда это всё взято))
Тут в комментарии написано /// Challenge model
При этом есть отдельно ChallengesModel
Ну и тд. Сделай нормальную конвенцию, типо сервис для бизнес логики, модель - это реально модель, ну и тд. Сейчас всякие провайдеры и тд - что это вообще

lankastersky replied:
переименовал ChallengesModel->ChallengesManager

NikAshanin commented on zen/models/Challenge.swift:
```diff
@@ -1,9 +1,86 @@
+
+    let prevStatuses: [Int]? = nil
+    let prevFinishedTimes: [TimeInterval]? = nil
+    let prevRatings: [Float]? = nil
```
оно и так nil, если ? ставишь
лишний код просто, удаляй

lankastersky replied:
done

NikAshanin commented on zen/models/Challenge.swift:
```diff
@@ -1,9 +1,86 @@
+    private(set) var comments: String?
+
+    let prevStatuses: [Int]? = nil
+    let prevFinishedTimes: [TimeInterval]? = nil
```
оно и так nil, если ? ставишь
лишний код просто, удаляй

lankastersky replied:
done

NikAshanin commented on zen/models/Challenge.swift:
```diff
@@ -1,9 +1,86 @@
+    private(set) var rating: Double?
+    private(set) var comments: String?
+
+    let prevStatuses: [Int]? = nil
```
оно и так nil, если ? ставишь
лишний код просто, удаляй

lankastersky replied:
done

NikAshanin commented on zen/models/Challenge.swift:
```diff
@@ -1,9 +1,86 @@
+        if let challengeLevel = ChallengeLevel(rawValue: levelInt) {
+            level = challengeLevel
+        } else {
+            level = .low
```
wtf?) как это и зачем?)

lankastersky replied:
что именно? если ты прод кастомный конструктор, то дефолтный не подходит потому что мне не все поля надо декодировать

NikAshanin replied:
почему если не распрарсило, то ты делаешь .low?

lankastersky replied:
согласен, поменял на эксепшн

NikAshanin commented on zen/models/ChallengeData.swift:
```diff
@@ -0,0 +1,26 @@
+import Foundation
+
+/// Modifiable challenge data for serialization etc.
+final class ChallengeData {
```
еще и Data суффикс. конценцию наименований я бы ввел все-таки и отписал сюда, что и как ты называешь. сейчас просто солянка супчик

lankastersky replied:
renamed to MutableChallengeData

NikAshanin commented on zen/models/ChallengesModel.swift:
```diff
@@ -0,0 +1,316 @@
+        if challengesMap.isEmpty {
+            if let challenges = challengesArchiver.restoreChallenges() {
+                challenges.forEach({ challenge in
+                    challengesMap[challenge.challengeId] = challenge
```
тут упадет

lankastersky replied:
почему? если такого ключа в мапе нет, просто вставка произойдет

NikAshanin commented on zen/views/ChallengeViewController.swift:
```diff
@@ -40,7 +41,8 @@ final class ChallengeViewController: UIViewController {
         let challengeIndex = 0
-        let challenge = Array(challenges.values)[challengeIndex]
-        contentLabel.text = challenge.content
+        if let challenge = challengesModel?.sortedChallenges[challengeIndex] {
```
упадет тут

lankastersky replied:
// TODO: choose proper index

NikAshanin commented on zen/views/JournalViewController.swift:
```diff
@@ -44,7 +44,7 @@ extension JournalViewController: UICollectionViewDataSource {
         }
 
-        let challenge: Challenge = Array(challenges.values)[indexPath.item]
+        let challenge: Challenge = challenges[indexPath.item]
```
упадет тут

lankastersky replied:
почему? я ж знаю сколько там элементов и возвращаю это при создании листа

NikAshanin replied:
считается небезопасным кодом. пример: приходит другой разраб и добавляет еще одну ячейку как хедер и всё падает

lankastersky replied:
ок добавил assert

lankastersky committed:
Code review fixed

lankastersky committed:
Code review fixes

NikAshanin commented on zen/models/ChallengesManager.swift:
```diff
@@ -0,0 +1,322 @@
+    private static let proportionAcceptMediumLavelChallenge = 1.0 / 3
+    private static let proportionAcceptHighLevelChallenge = 2.0 / 3
+
+    private var challengesMap: [String: Challenge] = [:]
```
Map переименуй, пожалуйста) Код пишем для других людей все-таки, согласно общепринятым языковым конвенциям

lankastersky replied:
done

NikAshanin commented on zen/models/ChallengesManager.swift:
```diff
@@ -0,0 +1,322 @@
+    var currentChallenge: Challenge? {
+        get {
+            if let currentChallengeId = currentChallengeId {
+                return challengesMap[currentChallengeId]
```
упадет тут. Обращение может произойти по ключу, которого нет в словаре

lankastersky replied:
если ключа нет в словаре то вернется nil что ок потому что возвращаемое значение опциональное

NikAshanin commented on zen/models/ChallengesManager.swift:
```diff
@@ -0,0 +1,322 @@
+    }
+
+    func challenge(_ challengeId: String) -> Challenge? {
+        return challengesMap[challengeId]
```
упадет тут. Обращение может произойти по ключу, которого нет в словаре

lankastersky replied:
если ключа нет в словаре то вернется nil что ок потому что возвращаемое значение опциональное

NikAshanin commented on zen/models/ChallengesManager.swift:
```diff
@@ -0,0 +1,322 @@
+
+        var newChallengeRequired = false
+
+        if currentChallengeId == "" {
```
нельзя сравнивать с пустой строкой. Вообще id не может быть пустой строкой. либо nil, либо значение нормальное. это только запутывает

lankastersky replied:
да, забыл убрать поменял на nil

NikAshanin commented on zen/models/ChallengesManager.swift:
```diff
@@ -0,0 +1,322 @@
+                + challengesByStatus(.shown)
+            let filteredNonFinishedChallenges = challengesOfCurrentLevel(nonFinishedChallenges)
+
+            if !(filteredNonFinishedChallenges.isEmpty) {
```
guard здесь лучше

lankastersky replied:
не думаю, это вполне законный сценарий, а гарды - когда по идее не должно это условие срабатывать.

NikAshanin commented on zen/models/ChallengesManager.swift:
```diff
@@ -0,0 +1,322 @@
+                return challenge.challengeId
+            }
+            let declinedChallenges = challengesByStatus(.declined)
+            if !declinedChallenges.isEmpty
```
guard здесь лучше

lankastersky replied:
не думаю, это вполне законный сценарий, а гарды - когда по идее не должно это условие срабатывать.

lankastersky committed:
Naming convention:
1) ...ViewController - контроллер.
2) ...Service - класс, у которого мы можем создавать инстансы. Содержит в себе бизнес-логику.
3) ...Helper - класс, содержит в себе только статические методы хелперы
4) ... - класс, который получаем от api + ряд дополнительный полей с нашими вычислениями
